src/components/workflow/workflow.service.ts

```diff
@@ -0,0 +1,866 @@
+  // attachNotificationGroupToState
+  async attachNotificationGroup(session: ISession, input: GroupState): Promise<void>{
+    try{
+      await this.db
+        .query()
+        .match([
+          [
+            node('token', 'Token', {
+              active: true,
+              value: session.token
+            }),
+            relation('in', '', 'token', {
+              active: true,
+            }),
+            node('requestingUser', 'User', {
+              id: session.userId
+            }),
+            relation('in', '', 'member'),
+            node('sg', 'SecurityGroup', {
+              id: input.securityGroupId
+            }),
+          ],
+          [
+            node('state', 'State', {
+              id: input.stateId
+            })
+          ]
+        ])
+        .merge([
+          [
+            node('state'),
+            relation('out', '', 'notification', {
+              active: true,
+              onEnter: true
+            }),
+            node('sg')
+          ]
+        ])
+        .first();
+
+    } catch (e) {
+      this.logger.warning('could not attach security group to state', {
+        exception: e
+      });
+      throw e;
+    }
+  }
+
+  // removeNotificationGroupFromState
+  async removeNotificationGroup(session: ISession, input: GroupState): Promise<void>{
+    try{
+      await this.db
+        .query()
+        .match([
+          [
+            node('token', 'Token', {
+              active: true,
+              value: session.token
+            }),
+            relation('in', '', 'token', {
+              active: true,
+            }),
+            node('requestingUser', 'User', {
+              id: session.userId
+            }),
+            relation('in', '', 'member'),
+            node('sg', 'SecurityGroup', {
+              id: input.securityGroupId
+            })
+          ],
+          [
+            node('state', 'State', {
+              id: input.stateId
+            }),
+            relation('out', 'rel', 'notification'),
+            node('sg')
+          ]
+        ])
+        .detachDelete('rel')
+        .first();
+
+    } catch (e) {
+      this.logger.warning('could not remove security group from state', {
+        exception: e
+      });
+      throw e;
+    }
+  }
+
+  // changeCurrentStateInWorkflow
+  async changeCurrentState(session: ISession, input: ChangeState): Promise<CommentState>{
+    try{
+
+      await this.db
+        .query()
+        .match([
+
+        ])
+        .run();
+    } catch (e) {
+      this.logger.warning('could not change current state', {
+        exception: e
+      });
+      throw e;
+    }
+    throw new NotImplementedException();
+  }
+
+  // creates a relationship from one state to another
+  // later we can create an abstracted function that creates and attaches a state to another state
+  // addPossibleStateToState
+  async addPossibleState(session: ISession, input: PossibleState): Promise<void>{
```

Comment:
user needs to be admin of base node and admin of sg.

Reply:
I'm not sure what I was thinking, but I was wrong on this one. The requesting user needs to be an admin on the BaseNode and that's it. Security groups don't come into play here. Sorry for that. 

The clause for the createWorkflow is an example for this function as well:

```
        .match([
          [
            node('token', 'Token', {
              active: true,
              value: session.token
            }),
            relation('in', '', 'token', {
              active: true
            }),
            node('user', 'User'),
            relation('in', '', 'admin', {
              active: true
            }),
            node('baseNode', 'BaseNode', {
              id: input.baseNodeId
            }),
          ],
        ])
```